# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: dom/media/systemservices/android_video_capture/video_capture_android.cc
# Commit: 21569aa4bced
# Full Hash: 21569aa4bcedb149bb3d1f5424f9d1d14e0fe441
# Author: Dan Minor <dminor@mozilla.com>
# Date: 2020-07-03 03:56:55
# Description:
#   Bug 1646220 - Remove unlinkCapturer from VideoCaptureAndroid; r=ng
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D81374
# ==============================================================================

diff -r 885d2ba42467 -r 21569aa4bced dom/media/systemservices/android_video_capture/video_capture_android.cc
--- a/dom/media/systemservices/android_video_capture/video_capture_android.cc	Thu Jul 02 18:07:10 2020 +0000
+++ b/dom/media/systemservices/android_video_capture/video_capture_android.cc	Mon Jun 29 18:45:45 2020 +0000
@@ -143,6 +143,8 @@
 void VideoCaptureAndroid::OnIncomingFrame(rtc::scoped_refptr<I420Buffer> buffer,
                                           int32_t degrees,
                                           int64_t captureTime) {
+  rtc::CritScope cs(&_apiCs);
+
   VideoRotation rotation =
       (degrees <= 45 || degrees > 315)
           ? kVideoRotation_0
@@ -179,12 +181,11 @@
   AttachThreadScoped ats(g_jvm_capture);
   JNIEnv* env = ats.env();
   jmethodID ctor = env->GetMethodID(g_java_capturer_class, "<init>",
-                                    "(Ljava/lang/String;J)V");
+                                    "(Ljava/lang/String;)V");
   assert(ctor);
   jstring j_deviceName = env->NewStringUTF(_deviceUniqueId);
-  jlong j_this = reinterpret_cast<intptr_t>(this);
   _jCapturer = env->NewGlobalRef(
-      env->NewObject(g_java_capturer_class, ctor, j_deviceName, j_this));
+      env->NewObject(g_java_capturer_class, ctor, j_deviceName));
   assert(_jCapturer);
   return 0;
 }
@@ -194,12 +195,6 @@
   if (_captureStarted) StopCapture();
   AttachThreadScoped ats(g_jvm_capture);
   JNIEnv* env = ats.env();
-
-  // Avoid callbacks into ourself even if the above stopCapture fails.
-  jmethodID j_unlink =
-      env->GetMethodID(g_java_capturer_class, "unlinkCapturer", "()V");
-  env->CallVoidMethod(_jCapturer, j_unlink);
-
   env->DeleteGlobalRef(_jCapturer);
 }
 
@@ -230,10 +225,11 @@
   _apiCs.Leave();
 
   jmethodID j_start =
-      env->GetMethodID(g_java_capturer_class, "startCapture", "(IIII)Z");
+      env->GetMethodID(g_java_capturer_class, "startCapture", "(IIIIJ)Z");
   assert(j_start);
+  jlong j_this = reinterpret_cast<intptr_t>(this);
   bool started = env->CallBooleanMethod(_jCapturer, j_start, width, height,
-                                        min_mfps, max_mfps);
+                                        min_mfps, max_mfps, j_this);
   if (started) {
     rtc::CritScope cs(&_apiCs);
     _requestedCapability = capability;
